### PR TITLE
Developer documentation

### PR DESCRIPTION
**Pull request by teccheck**
A lot of repositories on GitHub have a `HACKING.md` file that gives insight into how the project works and how it's structured. I think this would be awesome to have to get new developers on board (I guess it would have helped me get started developing Uchu).

Now I'm really not an expert when it comes to the Uchu codebase and what I've written is just my understanding from what I have read. As well as that English is not my native language so there might be mistakes with content, grammar and spelling. I'm open to any suggestions.

With that said do you guys think such a documentation would make sense and what do you think of my draft?

### CODE REVIEW

File: HACKING.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+## Introduction
+LU is build with the [Entity–component–system](https://en.wikipedia.org/wiki/Entity_component_system) which you may know from the Unity Engine. In short that means everything that you see on screen (and more) is a GameObject on which Components can be attached. These components give the object functionality like rendering, physics and so on. 
+
+LU uses four different types of servers to communicate. These are: The authentication server (Uchu.Auth), which handles login and redirection to the character server (Uchu.Char). The character server then handles the character selection and redirection to the world server (Uchu.World). Now the world server is the server that you actually play on. It handles everything from movement to chat to missions. There can be multiple World servers running at once but one world server only serves one world. There can however be multiple world servers serving the same world. This happens if the maximum amount of players on a server is reached.
```
Comment (Jettford) [inline HACKING.md]: as far as the client is aware there are 3 types of servers. Authentication, World (Character and World) and Chat (Never used/obsolete)

File: HACKING.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+* **Uchu.Navigation** - Code for pathfinding (used by the world server)
+* **Uchu.Physics.Test**- Unit Tests for Uchu.Physics
+* **Uchu.Physics** - Code for all kinds of physics calculations (used by the world server)
+* **Uchu.Python** - Python interpreter for scripts
+* **Uchu.Sso** - Code for lcdr's [Single-Sign-On](https://github.com/lcdr/sso_auth)
+* **Uchu.StandardScripts** - C# vserions of server scripts (most of the original lua server scripts are not available)
+* **Uchu.World.Test** - Unit Test for Uchu.World
+* **Uchu.World** - Code for the world server
+
+If you want to contribute to Uchu most likely Uchu.Wolrd is where you want to make your changes as it contains most of the actual game logic.
```
Comment (Jettford) [inline HACKING.md]: Typo

File: HACKING.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+## Logging
```
Comment (Jettford) [inline HACKING.md:46]: It might be worth having more sections like this, for things like Game message handlers, Game message sending functions etc. Just generic functions used all over the code base

File: HACKING.md (hunk 1 above)
Comment (teccheck) [reply HACKING.md]: Chat is handled by the world server right? And what about the UGC Server? Zaop told me about that one

File: HACKING.md (hunk 2 above)
Comment (teccheck) [reply HACKING.md]: Fixed

File: HACKING.md (hunk 1 above)
Comment (Jettford) [reply HACKING.md]: Chat is handled by the world server technically, chat packets are sent in a redirect packet that the world server is supposed to hand to the master server to handle

Comment (Jettford) [reply HACKING.md]: and the UGC server isn't an actual RakNet server like the rest, it is just a webserver

File: HACKING.md (hunk 3 above)
Comment (teccheck) [reply HACKING.md:46]: Added some stuff. What do you think of it?

File: HACKING.md (hunk 1 above)
Comment (teccheck) [reply HACKING.md]: I updated the intro, what do you think?

**Comment by teccheck** [comment]
By the way Wincent created a [card](https://github.com/UchuServer/Uchu/projects/1#card-29646095) about this topic

File: HACKING.md (hunk 3 above)
Comment (legocaptainprice) [reply HACKING.md:46]: spelling mistake, its Uchu.World :D

Comment (teccheck) [reply HACKING.md:46]: Already fixed, but thank you :)

File: HACKING.md [hunk 4]
```diff
@@ -91,7 +91,7 @@ namespace Uchu.World.Handlers.GameMessages
 
 As you can see all handlers are inside a class that extends `HandlerGroup`. The actual handling functions are annotated with `PacketHandler` and should always contain two arguments: The message to handle and the player that the message came from.
 
-Uchu will automatically find the right handler for a certain Game Message and invoke it so just have to define it.
+Uchu will automatically find the right handler for a certain Game Message and invoke it so you just have to define it.
```
Comment (Jettford) [inline HACKING.md]: Should probably be
> Uchu will use C# assembly reflection to find the right handler for a certain Game Message and invoke it so you just have to define it.

File: HACKING.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
+## Introduction
+LU is build with the [Entity–component–system](https://en.wikipedia.org/wiki/Entity_component_system) which you may know from the Unity Engine. In short that means everything that you see on screen (and more) is a GameObject on which Components can be attached. These components give the object functionality like rendering, physics and so on. 
+
+LU used two different types of servers to communicate. These are: 
+* The authentication server (Uchu.Auth), which handles login and redirection to the world server. 
+* The world server, which handles two things
+  * Character selection (Uchu.Char)
+  * Main game (Uchu.World)
+
+**Note:** A world server in Uchu only handles one world but there can be multiple world servers in one Uchu instance. They can even serve the same map if one world server has reached the maximum amount of player.
```
Comment (Frozenreflex) [inline HACKING.md]: There's a few small typos here and there, replace 'player' with 'players'

File: HACKING.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            // Do stuff
+        }
+    }
+}
+```
+
+As you can see all handlers are inside a class that extends `HandlerGroup`. The actual handling functions are annotated with `PacketHandler` and should always contain two arguments: The message to handle and the player that the message came from.
+
+Uchu will use C# assembly reflection to find the right handler for a certain Game Message and invoke it so you just have to define it.
```
Comment (Frozenreflex) [inline HACKING.md:94]: Something similar is done for StandardScripts

File: HACKING.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+* The world server, which handles two things
+  * Character selection (Uchu.Char)
+  * Main game (Uchu.World)
+
+**Note:** A world server in Uchu only handles one world but there can be multiple world servers in one Uchu instance. They can even serve the same map if one world server has reached the maximum amount of player. 
+
+The address of the auth server is configured in the `boot.cfg` file in the installation directory of the client. All other servers can have any address because the previous server sends the address to the client.
+
+There were also two other server types that aren't implemented in Uchu. These are
+* The Chat server which was never used because the world sever handles chat messages
```
Comment (Frozenreflex) [inline HACKING.md]: Add a comma after 'server'

File: HACKING.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+* **Uchu.Core.Test** - Unit Tests for Uchu.Core
+* **Uchu.Core** - Contains stuff that is used all over the project, like logging, io, database access, etc
+* **Uchu.Instance** - The base for all servers that Uchu runs (like auth, char, world)
+* **Uchu.Master** - This contains the main process that starts and controlls all the servers
+* **Uchu.Navigation** - Code for pathfinding (used by the world server)
+* **Uchu.Physics.Test**- Unit Tests for Uchu.Physics
+* **Uchu.Physics** - Code for all kinds of physics calculations (used by the world server)
+* **Uchu.Python** - Python interpreter for scripts
+* **Uchu.Sso** - Code for lcdr's [Single-Sign-On](https://github.com/lcdr/sso_auth)
+* **Uchu.StandardScripts** - C# versions of server scripts (most of the original lua server scripts are not available)
```
Comment (Frozenreflex) [inline HACKING.md]: I'd reword this to say "C# re-implementations of lua server scripts (most of which are not available in the client)"

File: HACKING.md (hunk 7 above)
Comment (Frozenreflex) [reply HACKING.md]: I'd maybe also reword this to say "The Chat server, which isn't used, since chat messages are handled in the world server."

File: HACKING.md (hunk 5 above)
Comment (teccheck) [reply HACKING.md]: Interesting typo. Imagine the server being like: "Nah that's too much player for today gonna make a new world server"

File: HACKING.md (hunk 7 above)
Comment (teccheck) [reply HACKING.md]: I never really understood where to place commas in English

File: HACKING.md (hunk 6 above)
Comment (teccheck) [reply HACKING.md:94]: Hm, I have no experience with StandardScripts so I can't really write anything about it

File: HACKING.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
+## Introduction
+LU is build with the [Entity–component–system](https://en.wikipedia.org/wiki/Entity_component_system) which you may know from the Unity Engine. In short that means everything that you see on screen (and more) is a GameObject on which Components can be attached. These components give the object functionality like rendering, physics and so on.
```
Comment (Jettford) [inline HACKING.md:2]: This is the wrong link I think, we don't use entity-component-system we use a entity-component system.

ECS contains an extra feature on top of EC that we and LU don't make use of.

File: HACKING.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+There are 3 submodules (marked with a blue text color in GitHub) in the repo.
+* [InfectedRose](https://github.com/Wincent01/InfectedRose) - A library to work with LEGO® Universe files
+* [NexusLogging](https://github.com/TheNexusAvenger/Nexus-Logging) - A logging library for .NET
+* [RakDotNet](https://github.com/UchuServer/RakDotNet) - A library for the RakNet protocol that LU uses
+
+Then there are also the folders that contain the actual Uchu code. These are:
+
+* **Uchu.Api** - It contains code for an api with which the Uchu instance can be controlled. There is a [Cli Tool](https://github.com/UchuServer/cli) that uses this api
+* **Uchu.Auth** - This contains the auth server
```
Comment (Jettford) [inline HACKING.md]: These should probably be clearer because they don't actually contain the auth server but rather the packet handlers and packets that Uchu.Instance use for the auth server. (Similar should go for Uchu.Char and Uchu.World)

**Comment by Jettford** [comment]
This is most looking quite good by now 👍

File: HACKING.md (hunk 1 above)
Comment (Jettford) [reply HACKING.md]: Seems good

File: HACKING.md (hunk 9 above)
Comment (teccheck) [reply HACKING.md:2]: I'm a bit confused about that. I thought that this ECS is just an architecural pattern

File: HACKING.md (hunk 10 above)
Comment (teccheck) [reply HACKING.md]: Tried to make it clearer. Is that better?

File: HACKING.md (hunk 6 above)
Comment (teccheck) [reply HACKING.md:94]: I added a small paragraph. What do you think of it?

File: HACKING.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+* `GetAllComponents()`
+* `TryGetComponent(Type type, out Component result)`
+* `TryGetComponents(Type type, out Component[] result)`
+* `TryGetComponent<T>(out T result)`
+* `TryGetComponents<T>(out T[] result)`
+
+The TryGet functions return a bool that says if a component of the requested type is present. Every component has a reference to it's GameObject, called `GameObject` that you can use to get access to other components.
+
+## Standard Scripts
+As mentioned before Uchu reimplements many of the server-side lua scripts in c# because most of them aren't included in the client and so are not accessible to the public. Similar to packet handlers the classes that represent those scripts have a special annotation that tells Uchu which script it replaces. Uchu will then, based on these annotations, find the right c# script to apply to a given object.
```
Comment (enteryournamehere) [inline HACKING.md]: The server scripts not being publicly available isn't entirely the reason for rewriting them in C#, it's also just that it is a much easier option than building something that can interface with Lua & implements all of the methods originally used in those scripts

**Comment by enteryournamehere** [comment]
Nice work, it's great to have this. Might be good to also add a link to this document in README.md.

**Comment by teccheck** [comment]
Should it be in this PR?

**Comment by enteryournamehere** [comment]
Yeah, you can just add a note about it to the Contributing section at the bottom

File: HACKING.md (hunk 11 above)
Comment (teccheck) [reply HACKING.md]: Is that better?

Comment (enteryournamehere) [reply HACKING.md]: Looks good to me :+1:

**Review by enteryournamehere — APPROVED** [review]

**Review by Frozenreflex — APPROVED** [review]
